File: javascript/manipulate.js
/**
 * Generate all unique manipulate of a given string.
 * @param {string} str - The input string
 * @return {string[]} - Array of unique manipulate
 */
function manipulate(str) {
  // Base case: if string is 1 character or less, return it
  if (str.length <= 1) return [str];

  let result = [];

  // Loop through each character in the string
  for (let i = 0; i < str.length; i++) {
    // Get the current character
    const currentChar = str[i];
    // Get the rest of the string without the current character
    const remainingChars = str.slice(0, i) + str.slice(i + 1);

    // Recursively get manipulate of the remaining characters
    const subManipulate = manipulate(remainingChars);

    // Add the current character to the beginning of each sub-permutation
    for (let subPerm of subManipulate) {
      result.push(currentChar + subPerm);
    }
  }

  // Remove duplicates by converting to Set and back to array
  return Array.from(new Set(result));
}

module.exports = manipulate;

File: golang/manipulate.go
package main

import "sort"

// manipulate generates all unique manipulations of a given string.
func manipulate(str string) []string {
	// Base case: if string is 1 character or less, return it
	if len(str) <= 1 {
		return []string{str}
	}

	var result []string

	// Loop through each character in the string
	for i := 0; i < len(str); i++ {
		// Get the current character
		currentChar := string(str[i])
		// Get the rest of the string without the current character
		remainingChars := str[:i] + str[i+1:]

		// Recursively get manipulations of the remaining characters
		subManipulations := manipulate(remainingChars)

		// Add the current character to the beginning of each sub-manipulation
		for _, subManip := range subManipulations {
			result = append(result, currentChar+subManip)
		}
	}

	// Remove duplicates
	uniqueResult := removeDuplicates(result)

	// Sort the result for consistency
	sort.Strings(uniqueResult)

	return uniqueResult
}

// removeDuplicates removes duplicate strings from a slice
func removeDuplicates(strSlice []string) []string {
	keys := make(map[string]bool)
	var list []string
	for _, entry := range strSlice {
		if _, value := keys[entry]; !value {
			keys[entry] = true
			list = append(list, entry)
		}
	}
	return list
}
